src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -89,20 +97,49 @@ class CoreWorkerDirectActorTaskSubmitter {
   void SetCallerCreationTimestamp(int64_t timestamp);
 
  private:
+  struct ClientQueue {
+    /// The current state of the actor. If this is ALIVE, then we should have
+    /// an RPC client to the actor. If this is DEAD, then all tasks in the
+    /// queue will be marked failed and all other ClientQueue state is ignored.
+    rpc::ActorTableData::ActorState state = rpc::ActorTableData::PENDING;
+    /// The RPC client. We use shared_ptr to enable shared_from_this for
+    /// pending client callbacks.
+    std::shared_ptr<rpc::CoreWorkerClientInterface> rpc_client = nullptr;
+    /// The intended worker ID of the actor.
+    std::string worker_id = "";
+
+    /// The actor's pending requests, ordered by the task number in the
+    /// request.
+    std::map<uint64_t, TaskSpecification> requests;
+    /// The send position of the next task to send to this actor. This sequence
+    /// number increases monotonically.
+    uint64_t next_send_position = 0;
+    /// The offset at which the the actor should start its counter for this
+    /// caller. This is used for actors that can be restarted, so that the new
+    /// instance of the actor knows from which task to start executing. This
+    /// increases monotonically.
+    uint64_t caller_starts_at = 0;
+    /// The number of tasks sent to this actor by this worker that will never
+    /// be submitted again. This is used to reset caller_starts_at if the actor
+    /// dies and is restarted.
```

Comment:
Consider adding an example ASCII, e.g., (note that I don't know if the diagram below is actually accurate)
```
1 2 3 4 5 6 7 8 9 10
                  ^ next send position
      ^ num_completed_tasks
^ caller starts at

Suppose we crash and recover, then tasks starting from 4 are resubmitted
1 2 3 4 5 6 7 8 9 10
                  ^ next send position
      ^ num_completed_tasks
      ^ caller starts at


```

Reply:
(I always find it hard to remember how this queuing protocol works after coming back from it for a long time. Would be good to write down the examples while you remember :) )